app/src/components/templateEditor/language/languageUtils.ts

```diff
@@ -0,0 +1,86 @@
+import { LRParser } from "@lezer/lr";
+
+/**
+ * Extracts all variables from a templated string.
+ *
+ * @param parser - The parser for the templating language.
+ *  The parser should be a language parser that emits Variable nodes.
+ * @param text - The text to extract variables from.
+ *
+ * @returns An array of variable names.
+ */
+export const extractVariables = (parser: LRParser, text: string) => {
+  const tree = parser.parse(text);
+  const variables: string[] = [];
+  const cur = tree.cursor();
+  do {
+    if (cur.name === "Variable") {
+      variables.push(text.slice(cur.node.from, cur.node.to));
+    }
+  } while (cur.next());
```

Comment:
any reason for a do ... while loop here? no big deal just not used very often so a bit harder to read

Reply:
It was recommended by the lezer docs, I can drop a link. It's because the tree cursor starts at the first node and we only want to iterate after the first loop.